internal/dnfjson/cache.go

```diff
@@ -1,46 +1,187 @@
 package dnfjson
 
 import (
+	"fmt"
 	"io/fs"
 	"os"
 	"path/filepath"
+	"sort"
+	"time"
+
+	"github.com/gobwas/glob"
 )
 
 type rpmCache struct {
 	// root path for the cache
 	root string
 
+	// individual repository cache data
+	repoElements map[string]pathInfo
+
+	// list of known repository IDs, sorted by mtime
+	repoRecency []string
+
+	// total cache size
+	size uint64
+
 	// max cache size
 	maxSize uint64
 }
 
 func newRPMCache(path string, maxSize uint64) *rpmCache {
-	return &rpmCache{
-		root:    path,
-		maxSize: maxSize,
+	r := &rpmCache{
+		root:         path,
+		repoElements: make(map[string]pathInfo),
+		size:         0,
+		maxSize:      maxSize,
 	}
+	// collect existing cache paths and timestamps
+	r.updateInfo()
+	return r
 }
 
-func (r *rpmCache) size() (uint64, error) {
-	var size uint64
-	sizer := func(path string, info fs.FileInfo, err error) error {
+func (r *rpmCache) clean() error {
+	curSize, err := dirSize(r.root)
+	if err != nil {
+		return err
+	}
+	if curSize > r.maxSize {
+		return os.RemoveAll(r.root)
+	}
+	return nil
+}
+
+// updateInfo updates the repoPaths and repoRecency fields of the rpmCache.
+func (r *rpmCache) updateInfo() {
+	repos := make(map[string]pathInfo)
+	repoIDs := make([]string, 0)
+	cacheEntries, _ := os.ReadDir(r.root)
+
+	var totalSize uint64
+
+	// collect the paths grouped by their repo ID (first 64 characters of a file or directory name)
```

Comment:
Could you explain why first 64? ideally in the comment.

Reply:
For repo IDs we use a sha256 of the repository config, which is 64 (hex) chars.  So we can just get the ID based on that assumption.  Another option would be to split the name on `-` and get the first element, in case we ever change the ID, but that might not work for everything.  For example the `.solv` file for a repo is `<ID>.solv`.

Here's an example cache with two repositories:
```
6e791d3334530bbed72986752494e81cc21ede63c4fee9f2b39ef25b879b1b3c-0d0eb10b0c459614/
6e791d3334530bbed72986752494e81cc21ede63c4fee9f2b39ef25b879b1b3c-filenames.solvx
6e791d3334530bbed72986752494e81cc21ede63c4fee9f2b39ef25b879b1b3c.solv
7e7c26ea9923d3a82e7a6049ac567c246e4ee0dcedb9183e01bf8c86e16e66d9-e58e3c9673d413e5/
7e7c26ea9923d3a82e7a6049ac567c246e4ee0dcedb9183e01bf8c86e16e66d9-filenames.solvx
7e7c26ea9923d3a82e7a6049ac567c246e4ee0dcedb9183e01bf8c86e16e66d9.solv
fastestmirror.cache
```

I'm open to suggestions for more robust ways of identifying repo IDs.

Reply:
Added comment.